### PR TITLE
add ViT as classification model TF and PT

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- adds ViT as classification model, which can be used as backbone for ViTSTR/ParSeq

Any feedback or ideas for improvements are very welcome :hugs: 

validated via test runs both models trains well

### CODE REVIEW

**Comment by felixdittrich92** [comment]
PT:
Epoch 1/10 - Validation loss: 2.14321 (Acc: 38.45%)
Validation loss decreased 2.14321 --> 1.93513: saving state...                                                                                                  
Epoch 2/10 - Validation loss: 1.93513 (Acc: 44.40%)

File: doctr/models/modules/vision_transformer/tensorflow.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+class PatchEmbedding(layers.Layer, NestedObject):
+    """Compute 2D patch embeddings with cls token and positional encoding"""
+
+    def __init__(
+        self,
+        img_size: Tuple[int, int],
+        patch_size: Tuple[int, int],
+        embed_dim: int,
+    ) -> None:
```
Comment (odulcy-mindee) [inline doctr/models/modules/vision_transformer/tensorflow.py:24]: `PatchEmbedding` signature in TensorFlow and PyTorch are not identical: there is `channels` kwarg in PyTorch for Conv2D but it's not needed for TensorFlow.

To have a common signature, maybe you can pass the whole `input_shape` (i.e. Image Size + Channels) in both TensorFlow and PyTorch implementation, and add `input_shape` as kwarg in `layers.Conv2D` in Tensorflow.

I must admit it's not needed, it's more a "cosmetic" suggestion.

File: doctr/models/modules/vision_transformer/tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        B, H, W, C = x.shape
+        assert H % self.patch_size[0] == 0, "Image height must be divisible by patch height"
+        assert W % self.patch_size[1] == 0, "Image width must be divisible by patch width"
+
+        patches = self.proj(x, **kwargs)  # BHWC
+
+        B, H, W, C = patches.shape
+        patches = tf.reshape(patches, (B, (H * W), C))  # (batch_size, num_patches, d_model)
+
+        cls_tokens = tf.repeat(self.cls_token, B, axis=0)  # (batch_size, num_patches, d_model)
```
Comment (odulcy-mindee) [inline doctr/models/modules/vision_transformer/tensorflow.py:60]:
```suggestion
        cls_tokens = tf.repeat(self.cls_token, B, axis=0)  # (batch_size, 1, d_model)
```

File: doctr/models/classification/vit/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        self,
+        input_shape: Tuple[int, int, int],
+        patch_size: Tuple[int, int] = (4, 4),
+        d_model: int = 768,
+        num_layers: int = 12,
+        num_heads: int = 12,
+        dropout: float = 0.0,
+        num_classes: int = 1000,
+        include_top: bool = True,
+        cfg: Optional[Dict[str, Any]] = None,
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply doctr/models/classification/vit/pytorch.py:58]: Hi @odulcy-mindee :wave: we need this cfg in every model for onnx

File: doctr/models/modules/vision_transformer/tensorflow.py (hunk 2 above)
Comment (felixdittrich92) [reply doctr/models/modules/vision_transformer/tensorflow.py:60]: thanks :+1:

File: doctr/models/modules/vision_transformer/tensorflow.py (hunk 1 above)
Comment (felixdittrich92) [reply doctr/models/modules/vision_transformer/tensorflow.py:24]: yep lets change it :)

**Review by odulcy-mindee — APPROVED** [review]
Really cool, thanks @felixdittrich92 ! :clap: :rocket:

**Comment by felixdittrich92** [comment]
> Really cool, thanks @felixdittrich92 ! clap rocket

yeah now we can continue with ViTSTR and ParSeq :hugs:

**Comment by felixdittrich92** [comment]
@frgfm late review maybe ? :sweat_smile:

File: doctr/models/classification/vit/pytorch.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+
+        embeddings = self.patch_embedding(x)
+        encoded = self.encoder(embeddings)
+
+        if self.include_top:
+            # (batch_size, num_classes) cls token
+            return self.head(encoded[:, 0])
+
+        return encoded
```
Comment (frgfm) [inline doctr/models/classification/vit/pytorch.py:79]: This very much looks like a Sequential, I'd suggest switching to that :)

File: doctr/models/modules/transformer/pytorch.py [hunk 5]
```diff
@@ -97,6 +100,45 @@ def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor, m
...
+        )
+
+    def forward(self, x: torch.Tensor, mask: Optional[torch.Tensor] = None) -> torch.Tensor:
+
+        output = x
+
+        for i in range(self.num_layers):
+            normed_output = self.layer_norm(output)
+            output = output + self.dropout(self.attention[i](normed_output, normed_output, normed_output, mask))
+            normed_output = self.layer_norm(output)
```
Comment (frgfm) [inline doctr/models/modules/transformer/pytorch.py:135]: mmmh I'm not 100% sure, but isn't it supposed to be another layer_norm instance?
Otherwise I imagine that the distribution stats will be biased :thinking:

File: docs/source/modules/models.rst [hunk 6]
```diff
@@ -29,6 +29,8 @@ doctr.models.classification
 
 .. autofunction:: doctr.models.classification.magc_resnet31
 
+.. autofunction:: doctr.models.classification.vit
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply docs/source/modules/models.rst:32]: yep sound good to me :+1:

File: doctr/models/modules/transformer/pytorch.py (hunk 5 above)
Comment (felixdittrich92) [reply doctr/models/modules/transformer/pytorch.py:135]: vit

Encoder block is the same as in normal transformer only Gelu instead of Relu in MLP

Comment (felixdittrich92) [reply doctr/models/modules/transformer/pytorch.py:135]: Damn i got what you mean... I have to update this also in the decoder :/
